File: python/src/falcon_core/physics/units/symbol_unit.py
from __future__ import annotations
from typing import Any, List, Dict, Tuple, Optional
from falcon_core._capi.symbol_unit import SymbolUnit as _CSymbolUnit

class SymbolUnit:
    """Python wrapper for SymbolUnit."""

    def __init__(self, c_obj):
        self._c = c_obj

    @classmethod
    def _from_capi(cls, c_obj):
        if c_obj is None:
            return None
        return cls(c_obj)

    @classmethod
    def SymbolUnit_create_meter(cls, ) -> SymbolUnit:
        return cls(_CSymbolUnit.SymbolUnit_create_meter())

    @classmethod
    def SymbolUnit_create_kilogram(cls, ) -> SymbolUnit:
        return cls(_CSymbolUnit.SymbolUnit_create_kilogram())

    @classmethod
    def SymbolUnit_create_second(cls, ) -> SymbolUnit:
        return cls(_CSymbolUnit.SymbolUnit_create_second())

    @classmethod
    def SymbolUnit_create_ampere(cls, ) -> SymbolUnit:
        return cls(_CSymbolUnit.SymbolUnit_create_ampere())

    @classmethod
    def SymbolUnit_create_kelvin(cls, ) -> SymbolUnit:
        return cls(_CSymbolUnit.SymbolUnit_create_kelvin())

    @classmethod
    def SymbolUnit_create_mole(cls, ) -> SymbolUnit:
        return cls(_CSymbolUnit.SymbolUnit_create_mole())

    @classmethod
    def SymbolUnit_create_candela(cls, ) -> SymbolUnit:
        return cls(_CSymbolUnit.SymbolUnit_create_candela())

    @classmethod
    def SymbolUnit_create_hertz(cls, ) -> SymbolUnit:
        return cls(_CSymbolUnit.SymbolUnit_create_hertz())

    @classmethod
    def SymbolUnit_create_newton(cls, ) -> SymbolUnit:
        return cls(_CSymbolUnit.SymbolUnit_create_newton())

    @classmethod
    def SymbolUnit_create_pascal(cls, ) -> SymbolUnit:
        return cls(_CSymbolUnit.SymbolUnit_create_pascal())

    @classmethod
    def SymbolUnit_create_joule(cls, ) -> SymbolUnit:
        return cls(_CSymbolUnit.SymbolUnit_create_joule())

    @classmethod
    def SymbolUnit_create_watt(cls, ) -> SymbolUnit:
        return cls(_CSymbolUnit.SymbolUnit_create_watt())

    @classmethod
    def SymbolUnit_create_coulomb(cls, ) -> SymbolUnit:
        return cls(_CSymbolUnit.SymbolUnit_create_coulomb())

    @classmethod
    def SymbolUnit_create_volt(cls, ) -> SymbolUnit:
        return cls(_CSymbolUnit.SymbolUnit_create_volt())

    @classmethod
    def SymbolUnit_create_farad(cls, ) -> SymbolUnit:
        return cls(_CSymbolUnit.SymbolUnit_create_farad())

    @classmethod
    def SymbolUnit_create_ohm(cls, ) -> SymbolUnit:
        return cls(_CSymbolUnit.SymbolUnit_create_ohm())

    @classmethod
    def SymbolUnit_create_siemens(cls, ) -> SymbolUnit:
        return cls(_CSymbolUnit.SymbolUnit_create_siemens())

    @classmethod
    def SymbolUnit_create_weber(cls, ) -> SymbolUnit:
        return cls(_CSymbolUnit.SymbolUnit_create_weber())

    @classmethod
    def SymbolUnit_create_tesla(cls, ) -> SymbolUnit:
        return cls(_CSymbolUnit.SymbolUnit_create_tesla())

    @classmethod
    def SymbolUnit_create_henry(cls, ) -> SymbolUnit:
        return cls(_CSymbolUnit.SymbolUnit_create_henry())

    @classmethod
    def SymbolUnit_create_minute(cls, ) -> SymbolUnit:
        return cls(_CSymbolUnit.SymbolUnit_create_minute())

    @classmethod
    def SymbolUnit_create_hour(cls, ) -> SymbolUnit:
        return cls(_CSymbolUnit.SymbolUnit_create_hour())

    @classmethod
    def SymbolUnit_create_electronvolt(cls, ) -> SymbolUnit:
        return cls(_CSymbolUnit.SymbolUnit_create_electronvolt())

    @classmethod
    def SymbolUnit_create_celsius(cls, ) -> SymbolUnit:
        return cls(_CSymbolUnit.SymbolUnit_create_celsius())

    @classmethod
    def SymbolUnit_create_fahrenheit(cls, ) -> SymbolUnit:
        return cls(_CSymbolUnit.SymbolUnit_create_fahrenheit())

    @classmethod
    def SymbolUnit_create_dimensionless(cls, ) -> SymbolUnit:
        return cls(_CSymbolUnit.SymbolUnit_create_dimensionless())

    @classmethod
    def SymbolUnit_create_percent(cls, ) -> SymbolUnit:
        return cls(_CSymbolUnit.SymbolUnit_create_percent())

    @classmethod
    def SymbolUnit_create_radian(cls, ) -> SymbolUnit:
        return cls(_CSymbolUnit.SymbolUnit_create_radian())

    @classmethod
    def SymbolUnit_create_kilometer(cls, ) -> SymbolUnit:
        return cls(_CSymbolUnit.SymbolUnit_create_kilometer())

    @classmethod
    def SymbolUnit_create_millimeter(cls, ) -> SymbolUnit:
        return cls(_CSymbolUnit.SymbolUnit_create_millimeter())

    @classmethod
    def SymbolUnit_create_millivolt(cls, ) -> SymbolUnit:
        return cls(_CSymbolUnit.SymbolUnit_create_millivolt())

    @classmethod
    def SymbolUnit_create_kilovolt(cls, ) -> SymbolUnit:
        return cls(_CSymbolUnit.SymbolUnit_create_kilovolt())

    @classmethod
    def SymbolUnit_create_milliampere(cls, ) -> SymbolUnit:
        return cls(_CSymbolUnit.SymbolUnit_create_milliampere())

    @classmethod
    def SymbolUnit_create_microampere(cls, ) -> SymbolUnit:
        return cls(_CSymbolUnit.SymbolUnit_create_microampere())

    @classmethod
    def SymbolUnit_create_nanoampere(cls, ) -> SymbolUnit:
        return cls(_CSymbolUnit.SymbolUnit_create_nanoampere())

    @classmethod
    def SymbolUnit_create_picoampere(cls, ) -> SymbolUnit:
        return cls(_CSymbolUnit.SymbolUnit_create_picoampere())

    @classmethod
    def SymbolUnit_create_millisecond(cls, ) -> SymbolUnit:
        return cls(_CSymbolUnit.SymbolUnit_create_millisecond())

    @classmethod
    def SymbolUnit_create_microsecond(cls, ) -> SymbolUnit:
        return cls(_CSymbolUnit.SymbolUnit_create_microsecond())

    @classmethod
    def SymbolUnit_create_nanosecond(cls, ) -> SymbolUnit:
        return cls(_CSymbolUnit.SymbolUnit_create_nanosecond())

    @classmethod
    def SymbolUnit_create_picosecond(cls, ) -> SymbolUnit:
        return cls(_CSymbolUnit.SymbolUnit_create_picosecond())

    @classmethod
    def SymbolUnit_create_milliohm(cls, ) -> SymbolUnit:
        return cls(_CSymbolUnit.SymbolUnit_create_milliohm())

    @classmethod
    def SymbolUnit_create_kiloohm(cls, ) -> SymbolUnit:
        return cls(_CSymbolUnit.SymbolUnit_create_kiloohm())

    @classmethod
    def SymbolUnit_create_megaohm(cls, ) -> SymbolUnit:
        return cls(_CSymbolUnit.SymbolUnit_create_megaohm())

    @classmethod
    def SymbolUnit_create_millihertz(cls, ) -> SymbolUnit:
        return cls(_CSymbolUnit.SymbolUnit_create_millihertz())

    @classmethod
    def SymbolUnit_create_kilohertz(cls, ) -> SymbolUnit:
        return cls(_CSymbolUnit.SymbolUnit_create_kilohertz())

    @classmethod
    def SymbolUnit_create_megahertz(cls, ) -> SymbolUnit:
        return cls(_CSymbolUnit.SymbolUnit_create_megahertz())

    @classmethod
    def SymbolUnit_create_gigahertz(cls, ) -> SymbolUnit:
        return cls(_CSymbolUnit.SymbolUnit_create_gigahertz())

    @classmethod
    def SymbolUnit_create_meters_per_second(cls, ) -> SymbolUnit:
        return cls(_CSymbolUnit.SymbolUnit_create_meters_per_second())

    @classmethod
    def SymbolUnit_create_meters_per_second_squared(cls, ) -> SymbolUnit:
        return cls(_CSymbolUnit.SymbolUnit_create_meters_per_second_squared())

    @classmethod
    def SymbolUnit_create_newton_meter(cls, ) -> SymbolUnit:
        return cls(_CSymbolUnit.SymbolUnit_create_newton_meter())

    @classmethod
    def SymbolUnit_create_newtons_per_meter(cls, ) -> SymbolUnit:
        return cls(_CSymbolUnit.SymbolUnit_create_newtons_per_meter())

    @classmethod
    def SymbolUnit_create_volts_per_meter(cls, ) -> SymbolUnit:
        return cls(_CSymbolUnit.SymbolUnit_create_volts_per_meter())

    @classmethod
    def SymbolUnit_create_volts_per_second(cls, ) -> SymbolUnit:
        return cls(_CSymbolUnit.SymbolUnit_create_volts_per_second())

    @classmethod
    def SymbolUnit_create_amperes_per_meter(cls, ) -> SymbolUnit:
        return cls(_CSymbolUnit.SymbolUnit_create_amperes_per_meter())

    @classmethod
    def SymbolUnit_create_volts_per_ampere(cls, ) -> SymbolUnit:
        return cls(_CSymbolUnit.SymbolUnit_create_volts_per_ampere())

    @classmethod
    def SymbolUnit_create_watts_per_meter_kelvin(cls, ) -> SymbolUnit:
        return cls(_CSymbolUnit.SymbolUnit_create_watts_per_meter_kelvin())

    @classmethod
    def SymbolUnit_from_json_string(cls, json: str) -> SymbolUnit:
        return cls(_CSymbolUnit.SymbolUnit_from_json_string(json))

    def symbol(self, ) -> str:
        ret = self._c.symbol()
        return ret

    def name(self, ) -> str:
        ret = self._c.name()
        return ret

    def multiplication(self, other: SymbolUnit) -> SymbolUnit:
        ret = self._c.multiplication(other._c)
        return cls._from_capi(ret)

    def division(self, other: SymbolUnit) -> SymbolUnit:
        ret = self._c.division(other._c)
        return cls._from_capi(ret)

    def power(self, power: Any) -> SymbolUnit:
        ret = self._c.power(power)
        return cls._from_capi(ret)

    def with_prefix(self, prefix: str) -> SymbolUnit:
        ret = self._c.with_prefix(prefix)
        return cls._from_capi(ret)

    def convert_value_to(self, value: Any, target: SymbolUnit) -> None:
        ret = self._c.convert_value_to(value, target._c)
        return ret

    def is_compatible_with(self, other: SymbolUnit) -> None:
        ret = self._c.is_compatible_with(other._c)
        return ret

    def equal(self, other: SymbolUnit) -> None:
        ret = self._c.equal(other._c)
        return ret

    def not_equal(self, other: SymbolUnit) -> None:
        ret = self._c.not_equal(other._c)
        return ret

    def to_json_string(self, ) -> str:
        ret = self._c.to_json_string()
        return ret

    def __mul__(self, other):
        """Operator overload for *"""
        if isinstance(other, (int, float)):
            return self.multiplication(other)
        return NotImplemented

    def __truediv__(self, other):
        """Operator overload for /"""
        if isinstance(other, (int, float)):
            return self.division(other)
        return NotImplemented

    def __eq__(self, other):
        """Operator overload for =="""
        if not isinstance(other, SymbolUnit):
            return NotImplemented
        return self.equality(other)

    def __ne__(self, other):
        """Operator overload for !="""
        if not isinstance(other, SymbolUnit):
            return NotImplemented
        return self.notequality(other)

File: ocaml/_opam/.opam-switch/sources/falcon_core/src/falcon_core/physics/units/symbolunit.ml
open Ctypes
open Capi_bindings
open ErrorHandling

(* no extra imports *)

class c_symbolunit (h : unit ptr) = object(self)
  val raw_val = h
  method raw = raw_val
  initializer Gc.finalise (fun _ ->
    Capi_bindings.symbolunit_destroy raw_val;
    ErrorHandling.raise_if_error ()
  ) self
end

module SymbolUnit = struct
  type t = c_symbolunit

  let copy (handle : t) : t =
    ErrorHandling.read handle (fun () ->
      let ptr = Capi_bindings.symbolunit_copy handle#raw in
      ErrorHandling.raise_if_error ();
      new c_symbolunit ptr
    )

  let fromjson (json : string) : t =
    let ptr = Capi_bindings.symbolunit_from_json_string (Capi_bindings.string_wrap json) in
    ErrorHandling.raise_if_error ();
    new c_symbolunit ptr

  let meter () : t =
    let ptr = Capi_bindings.symbolunit_create_meter () in
    ErrorHandling.raise_if_error ();
    new c_symbolunit ptr

  let kilogram () : t =
    let ptr = Capi_bindings.symbolunit_create_kilogram () in
    ErrorHandling.raise_if_error ();
    new c_symbolunit ptr

  let second () : t =
    let ptr = Capi_bindings.symbolunit_create_second () in
    ErrorHandling.raise_if_error ();
    new c_symbolunit ptr

  let ampere () : t =
    let ptr = Capi_bindings.symbolunit_create_ampere () in
    ErrorHandling.raise_if_error ();
    new c_symbolunit ptr

  let kelvin () : t =
    let ptr = Capi_bindings.symbolunit_create_kelvin () in
    ErrorHandling.raise_if_error ();
    new c_symbolunit ptr

  let mole () : t =
    let ptr = Capi_bindings.symbolunit_create_mole () in
    ErrorHandling.raise_if_error ();
    new c_symbolunit ptr

  let candela () : t =
    let ptr = Capi_bindings.symbolunit_create_candela () in
    ErrorHandling.raise_if_error ();
    new c_symbolunit ptr

  let hertz () : t =
    let ptr = Capi_bindings.symbolunit_create_hertz () in
    ErrorHandling.raise_if_error ();
    new c_symbolunit ptr

  let newton () : t =
    let ptr = Capi_bindings.symbolunit_create_newton () in
    ErrorHandling.raise_if_error ();
    new c_symbolunit ptr

  let pascal () : t =
    let ptr = Capi_bindings.symbolunit_create_pascal () in
    ErrorHandling.raise_if_error ();
    new c_symbolunit ptr

  let joule () : t =
    let ptr = Capi_bindings.symbolunit_create_joule () in
    ErrorHandling.raise_if_error ();
    new c_symbolunit ptr

  let watt () : t =
    let ptr = Capi_bindings.symbolunit_create_watt () in
    ErrorHandling.raise_if_error ();
    new c_symbolunit ptr

  let coulomb () : t =
    let ptr = Capi_bindings.symbolunit_create_coulomb () in
    ErrorHandling.raise_if_error ();
    new c_symbolunit ptr

  let volt () : t =
    let ptr = Capi_bindings.symbolunit_create_volt () in
    ErrorHandling.raise_if_error ();
    new c_symbolunit ptr

  let farad () : t =
    let ptr = Capi_bindings.symbolunit_create_farad () in
    ErrorHandling.raise_if_error ();
    new c_symbolunit ptr

  let ohm () : t =
    let ptr = Capi_bindings.symbolunit_create_ohm () in
    ErrorHandling.raise_if_error ();
    new c_symbolunit ptr

  let siemens () : t =
    let ptr = Capi_bindings.symbolunit_create_siemens () in
    ErrorHandling.raise_if_error ();
    new c_symbolunit ptr

  let weber () : t =
    let ptr = Capi_bindings.symbolunit_create_weber () in
    ErrorHandling.raise_if_error ();
    new c_symbolunit ptr

  let tesla () : t =
    let ptr = Capi_bindings.symbolunit_create_tesla () in
    ErrorHandling.raise_if_error ();
    new c_symbolunit ptr

  let henry () : t =
    let ptr = Capi_bindings.symbolunit_create_henry () in
    ErrorHandling.raise_if_error ();
    new c_symbolunit ptr

  let minute () : t =
    let ptr = Capi_bindings.symbolunit_create_minute () in
    ErrorHandling.raise_if_error ();
    new c_symbolunit ptr

  let hour () : t =
    let ptr = Capi_bindings.symbolunit_create_hour () in
    ErrorHandling.raise_if_error ();
    new c_symbolunit ptr

  let electronvolt () : t =
    let ptr = Capi_bindings.symbolunit_create_electronvolt () in
    ErrorHandling.raise_if_error ();
    new c_symbolunit ptr

  let celsius () : t =
    let ptr = Capi_bindings.symbolunit_create_celsius () in
    ErrorHandling.raise_if_error ();
    new c_symbolunit ptr

  let fahrenheit () : t =
    let ptr = Capi_bindings.symbolunit_create_fahrenheit () in
    ErrorHandling.raise_if_error ();
    new c_symbolunit ptr

  let dimensionless () : t =
    let ptr = Capi_bindings.symbolunit_create_dimensionless () in
    ErrorHandling.raise_if_error ();
    new c_symbolunit ptr

  let percent () : t =
    let ptr = Capi_bindings.symbolunit_create_percent () in
    ErrorHandling.raise_if_error ();
    new c_symbolunit ptr

  let radian () : t =
    let ptr = Capi_bindings.symbolunit_create_radian () in
    ErrorHandling.raise_if_error ();
    new c_symbolunit ptr

  let kilometer () : t =
    let ptr = Capi_bindings.symbolunit_create_kilometer () in
    ErrorHandling.raise_if_error ();
    new c_symbolunit ptr

  let millimeter () : t =
    let ptr = Capi_bindings.symbolunit_create_millimeter () in
    ErrorHandling.raise_if_error ();
    new c_symbolunit ptr

  let millivolt () : t =
    let ptr = Capi_bindings.symbolunit_create_millivolt () in
    ErrorHandling.raise_if_error ();
    new c_symbolunit ptr

  let kilovolt () : t =
    let ptr = Capi_bindings.symbolunit_create_kilovolt () in
    ErrorHandling.raise_if_error ();
    new c_symbolunit ptr

  let milliampere () : t =
    let ptr = Capi_bindings.symbolunit_create_milliampere () in
    ErrorHandling.raise_if_error ();
    new c_symbolunit ptr

  let microampere () : t =
    let ptr = Capi_bindings.symbolunit_create_microampere () in
    ErrorHandling.raise_if_error ();
    new c_symbolunit ptr

  let nanoampere () : t =
    let ptr = Capi_bindings.symbolunit_create_nanoampere () in
    ErrorHandling.raise_if_error ();
    new c_symbolunit ptr

  let picoampere () : t =
    let ptr = Capi_bindings.symbolunit_create_picoampere () in
    ErrorHandling.raise_if_error ();
    new c_symbolunit ptr

  let millisecond () : t =
    let ptr = Capi_bindings.symbolunit_create_millisecond () in
    ErrorHandling.raise_if_error ();
    new c_symbolunit ptr

  let microsecond () : t =
    let ptr = Capi_bindings.symbolunit_create_microsecond () in
    ErrorHandling.raise_if_error ();
    new c_symbolunit ptr

  let nanosecond () : t =
    let ptr = Capi_bindings.symbolunit_create_nanosecond () in
    ErrorHandling.raise_if_error ();
    new c_symbolunit ptr

  let picosecond () : t =
    let ptr = Capi_bindings.symbolunit_create_picosecond () in
    ErrorHandling.raise_if_error ();
    new c_symbolunit ptr

  let milliohm () : t =
    let ptr = Capi_bindings.symbolunit_create_milliohm () in
    ErrorHandling.raise_if_error ();
    new c_symbolunit ptr

  let kiloohm () : t =
    let ptr = Capi_bindings.symbolunit_create_kiloohm () in
    ErrorHandling.raise_if_error ();
    new c_symbolunit ptr

  let megaohm () : t =
    let ptr = Capi_bindings.symbolunit_create_megaohm () in
    ErrorHandling.raise_if_error ();
    new c_symbolunit ptr

  let millihertz () : t =
    let ptr = Capi_bindings.symbolunit_create_millihertz () in
    ErrorHandling.raise_if_error ();
    new c_symbolunit ptr

  let kilohertz () : t =
    let ptr = Capi_bindings.symbolunit_create_kilohertz () in
    ErrorHandling.raise_if_error ();
    new c_symbolunit ptr

  let megahertz () : t =
    let ptr = Capi_bindings.symbolunit_create_megahertz () in
    ErrorHandling.raise_if_error ();
    new c_symbolunit ptr

  let gigahertz () : t =
    let ptr = Capi_bindings.symbolunit_create_gigahertz () in
    ErrorHandling.raise_if_error ();
    new c_symbolunit ptr

  let metersPerSecond () : t =
    let ptr = Capi_bindings.symbolunit_create_meters_per_second () in
    ErrorHandling.raise_if_error ();
    new c_symbolunit ptr

  let metersPerSecondSquared () : t =
    let ptr = Capi_bindings.symbolunit_create_meters_per_second_squared () in
    ErrorHandling.raise_if_error ();
    new c_symbolunit ptr

  let newtonMeter () : t =
    let ptr = Capi_bindings.symbolunit_create_newton_meter () in
    ErrorHandling.raise_if_error ();
    new c_symbolunit ptr

  let newtonsPerMeter () : t =
    let ptr = Capi_bindings.symbolunit_create_newtons_per_meter () in
    ErrorHandling.raise_if_error ();
    new c_symbolunit ptr

  let voltsPerMeter () : t =
    let ptr = Capi_bindings.symbolunit_create_volts_per_meter () in
    ErrorHandling.raise_if_error ();
    new c_symbolunit ptr

  let voltsPerSecond () : t =
    let ptr = Capi_bindings.symbolunit_create_volts_per_second () in
    ErrorHandling.raise_if_error ();
    new c_symbolunit ptr

  let amperesPerMeter () : t =
    let ptr = Capi_bindings.symbolunit_create_amperes_per_meter () in
    ErrorHandling.raise_if_error ();
    new c_symbolunit ptr

  let voltsPerAmpere () : t =
    let ptr = Capi_bindings.symbolunit_create_volts_per_ampere () in
    ErrorHandling.raise_if_error ();
    new c_symbolunit ptr

  let wattsPerMeterKelvin () : t =
    let ptr = Capi_bindings.symbolunit_create_watts_per_meter_kelvin () in
    ErrorHandling.raise_if_error ();
    new c_symbolunit ptr

  let equal (handle : t) (other : t) : bool =
    ErrorHandling.multi_read [handle; other] (fun () ->
      let result = Capi_bindings.symbolunit_equal handle#raw other#raw in
      ErrorHandling.raise_if_error ();
      result
    )

  let notEqual (handle : t) (other : t) : bool =
    ErrorHandling.multi_read [handle; other] (fun () ->
      let result = Capi_bindings.symbolunit_not_equal handle#raw other#raw in
      ErrorHandling.raise_if_error ();
      result
    )

  let toJsonString (handle : t) : string =
    ErrorHandling.read handle (fun () ->
      let result = Capi_bindings.symbolunit_to_json_string handle#raw in
      ErrorHandling.raise_if_error ();
      Capi_bindings.string_to_ocaml result
    )

  let symbol (handle : t) : string =
    ErrorHandling.read handle (fun () ->
      let result = Capi_bindings.symbolunit_symbol handle#raw in
      ErrorHandling.raise_if_error ();
      Capi_bindings.string_to_ocaml result
    )

  let name (handle : t) : string =
    ErrorHandling.read handle (fun () ->
      let result = Capi_bindings.symbolunit_name handle#raw in
      ErrorHandling.raise_if_error ();
      Capi_bindings.string_to_ocaml result
    )

  let multiplication (handle : t) (other : t) : t =
    ErrorHandling.multi_read [handle; other] (fun () ->
      let result = Capi_bindings.symbolunit_multiplication handle#raw other#raw in
      ErrorHandling.raise_if_error ();
      new c_symbolunit result
    )

  let division (handle : t) (other : t) : t =
    ErrorHandling.multi_read [handle; other] (fun () ->
      let result = Capi_bindings.symbolunit_division handle#raw other#raw in
      ErrorHandling.raise_if_error ();
      new c_symbolunit result
    )

  let power (handle : t) (power : int) : t =
    ErrorHandling.read handle (fun () ->
      let result = Capi_bindings.symbolunit_power handle#raw power in
      ErrorHandling.raise_if_error ();
      new c_symbolunit result
    )

  let withPrefix (handle : t) (prefix : string) : t =
    ErrorHandling.read handle (fun () ->
      let result = Capi_bindings.symbolunit_with_prefix handle#raw (Capi_bindings.string_wrap prefix) in
      ErrorHandling.raise_if_error ();
      new c_symbolunit result
    )

  let convertValueTo (handle : t) (value : float) (target : t) : float =
    ErrorHandling.multi_read [handle; target] (fun () ->
      let result = Capi_bindings.symbolunit_convert_value_to handle#raw value target#raw in
      ErrorHandling.raise_if_error ();
      result
    )

  let isCompatibleWith (handle : t) (other : t) : bool =
    ErrorHandling.multi_read [handle; other] (fun () ->
      let result = Capi_bindings.symbolunit_is_compatible_with handle#raw other#raw in
      ErrorHandling.raise_if_error ();
      result
    )

end